src/Microsoft.ML.TimeSeries/PredictionEngine.cs

```diff
@@ -0,0 +1,207 @@
+            : base(env, transformer, ignoreMissingColumns, inputSchemaDefinition, outputSchemaDefinition)
+        {
+        }
+
+        public void GetStatefulRows(IRow input, IRowToRowMapper mapper, Func<int, bool> active,
+            List<IStatefulRow> rows, out Action disposer, out IRow outRow)
+        {
+            Contracts.CheckValue(input, nameof(input));
+            Contracts.CheckValue(active, nameof(active));
+
+            disposer = null;
+            IRowToRowMapper[] innerMappers = new IRowToRowMapper[0];
+            if (mapper is CompositeRowToRowMapper)
+                innerMappers = ((CompositeRowToRowMapper)mapper).InnerMappers;
+
+            if (innerMappers.Length == 0)
+            {
+                bool differentActive = false;
+                for (int c = 0; c < input.Schema.ColumnCount; ++c)
+                {
+                    bool wantsActive = active(c);
+                    bool isActive = input.IsColumnActive(c);
+                    differentActive |= wantsActive != isActive;
+
+                    if (wantsActive && !isActive)
+                        throw Contracts.ExceptParam(nameof(input), $"Mapper required column '{input.Schema.GetColumnName(c)}' active but it was not.");
+                }
+
+                outRow = input;
+                return;
+            }
+
+            // For each of the inner mappers, we will be calling their GetRow method, but to do so we need to know
+            // what we need from them. The last one will just have the input, but the rest will need to be
+            // computed based on the dependencies of the next one in the chain.
+            var deps = new Func<int, bool>[innerMappers.Length];
+            deps[deps.Length - 1] = active;
+            for (int i = deps.Length - 1; i >= 1; --i)
+                deps[i - 1] = innerMappers[i].GetDependencies(deps[i]);
+
+            IRow result = input;
+            for (int i = 0; i < innerMappers.Length; ++i)
+            {
+                Action localDisp;
+                if (innerMappers[i] is CompositeRowToRowMapper)
+                    GetStatefulRows(result, innerMappers[i], deps[i], rows, out localDisp, out result);
+                else
+                    result = innerMappers[i].GetRow(result, deps[i], out localDisp);
+
+                if (result is IStatefulRow)
+                    rows.Add((IStatefulRow)result);
+
+                if (localDisp != null)
+                {
+                    if (disposer == null)
+                        disposer = localDisp;
+                    else
+                        disposer = localDisp + disposer;
+                    // We want the last disposer to be called first, so the order of the addition here is important.
+                }
+            }
+
+            outRow = result;
+        }
+
+        private ValuePinger[][] CreatePingers(List<IStatefulRow> rows)
+        {
```

Comment:
>{ [](start = 8, length = 1)

I think you can just merge them all onto one delegate, there's no need to have arrays of arrays of delegates